freqtrade/data/dataprovider.py

```diff
@@ -63,9 +80,112 @@ def _set_cached_df(
         :param dataframe: analyzed dataframe
         :param candle_type: Any of the enum CandleType (must match trading mode!)
         """
-        self.__cached_pairs[(pair, timeframe, candle_type)] = (
+        pair_key = (pair, timeframe, candle_type)
+        self.__cached_pairs[pair_key] = (
             dataframe, datetime.now(timezone.utc))
 
+    # For multiple producers we will want to merge the pairlists instead of overwriting
+    def _set_producer_pairs(self, pairlist: List[str], producer_name: str = "default"):
+        """
+        Set the pairs received to later be used.
+
+        :param pairlist: List of pairs
+        """
+        self.__producer_pairs[producer_name] = pairlist
+
+    def get_producer_pairs(self, producer_name: str = "default") -> List[str]:
+        """
+        Get the pairs cached from the producer
+
+        :returns: List of pairs
+        """
+        return self.__producer_pairs.get(producer_name, []).copy()
+
+    def _emit_df(
+        self,
+        pair_key: PairWithTimeframe,
+        dataframe: DataFrame
+    ) -> None:
+        """
+        Send this dataframe as an ANALYZED_DF message to RPC
+
+        :param pair_key: PairWithTimeframe tuple
+        :param data: Tuple containing the DataFrame and the datetime it was cached
+        """
+        if self.__rpc:
+            self.__rpc.send_msg(
+                {
+                    'type': RPCMessageType.ANALYZED_DF,
+                    'data': {
+                        'key': pair_key,
+                        'df': dataframe,
+                        'la': datetime.now(timezone.utc)
+                    }
+                }
+            )
+
+    def _add_external_df(
+        self,
+        pair: str,
+        dataframe: DataFrame,
+        last_analyzed: Optional[datetime] = None,
```

Comment:
having `last_analyzed` optional (with default to None) seems odd - as we control the only occurance of this - and that code requires last_analyzed to be set.
Is there some deeper thought about potential future usecases behind, or just copy/paste from the getter?

The same is true for `timeframe` - and actually also for `candle_type` - all are known - and required for us to actually store a dataframe. 
We can't assume a "default timeframe" - as the leader could be using a diff. timeframe than the follower.

The only parameter that might be optional is the producer_name - but here, the "producing" code defaults to "default" already - so i think it'd be cleaner for the interface (as well as for the code within) to require all parameters to be set.

Obviously that ONLY applies to the `_add_` method - for the getter, having defaults is perfectly fine.

Reply:
I believe this was just left over from before we made the changes to sending the last_analyzed with the df on the leader side. I will change these.